logs/stream/stream.go

```diff
@@ -44,41 +44,81 @@ import (
 func findReadyLogLines(logLines []state.LogLine, threshold time.Duration) ([]state.LogLine, []state.LogLine) {
 	var readyLogLines []state.LogLine
 	var tooFreshLogLines []state.LogLine
-	var lastPrimaryLogLine *state.LogLine
+	var lastLogLineWithPrefixPid int32 = -1
+	var lastMainLogLinePid int32 = -1
 
 	now := time.Now()
 
 	for _, logLine := range logLines {
+		ready := false
 		if logLine.LogLevel == pganalyze_collector.LogLineInformation_UNKNOWN {
 			if logLine.BackendPid != 0 {
 				// Heroku case (Unknown log lines have PIDs assigned)
-				if lastPrimaryLogLine != nil && logLine.BackendPid == lastPrimaryLogLine.BackendPid {
-					readyLogLines = append(readyLogLines, logLine)
-				} else {
-					tooFreshLogLines = append(tooFreshLogLines, logLine)
-				}
+				ready = lastLogLineWithPrefixPid != -1 && logLine.BackendPid == lastLogLineWithPrefixPid
 			} else {
-				if lastPrimaryLogLine != nil {
-					// Self-managed and Google Cloud SQL case - we always stitch unknown log levels to the prior line
-					readyLogLines = append(readyLogLines, logLine)
-				} else if now.Sub(logLine.CollectedAt) <= threshold {
-					tooFreshLogLines = append(tooFreshLogLines, logLine)
+				// Self-managed case (subsequent multi-line lines do not have PIDs)
+				if lastLogLineWithPrefixPid != -1 {
+					// We always stitch unknown log levels to the prior line
+					ready = true
+				} else if now.Sub(logLine.CollectedAt) > threshold {
+					// This is ready, but we can't associate it to anything within the threshold - throw it away
+					continue
 				}
 			}
 		} else {
 			if now.Sub(logLine.CollectedAt) > threshold {
-				readyLogLines = append(readyLogLines, logLine)
-				lastPrimaryLogLine = &logLine
+				ready = true
+				lastLogLineWithPrefixPid = logLine.BackendPid
+				if isPostgresErrorLevel(logLine.LogLevel) {
+					lastMainLogLinePid = logLine.BackendPid
+				} else {
+					lastMainLogLinePid = -1
+				}
+			} else if isAdditionalLineLevel(logLine.LogLevel) && logLine.BackendPid != 0 && lastMainLogLinePid != -1 && logLine.BackendPid == lastMainLogLinePid {
+				ready = true
+				lastLogLineWithPrefixPid = logLine.BackendPid
 			} else {
-				tooFreshLogLines = append(tooFreshLogLines, logLine)
-				lastPrimaryLogLine = nil
+				lastLogLineWithPrefixPid = -1
```

Comment:
Might be worth commenting this branch? This is the case for a too-fresh main line, right?

Reply:
Yeah, good point - adding a comment.